src/model/schema.js

```diff
@@ -668,35 +668,60 @@ export default class Schema {
 	 * @returns {Array.<module:engine/model/range~Range>} Ranges in which the attribute is allowed.
 	 */
 	getValidRanges( ranges, attribute ) {
-		const validRanges = [];
+		ranges = convertToMinimalFlatRanges( ranges );
+		const result = [];
 
 		for ( const range of ranges ) {
-			let last = range.start;
-			let from = range.start;
-			const to = range.end;
-
-			for ( const value of range.getWalker() ) {
-				if ( !this.checkAttribute( value.item, attribute ) ) {
-					if ( !from.isEqual( last ) ) {
-						validRanges.push( new Range( from, last ) );
-					}
+			const validRanges = this._getValidRangesForRange( range, attribute );
```

Comment:
Looks like a great case for a generator :)

Reply:
Is there any gain by here by doing that (other than being cool)?

Reply:
Less code. Also, I think it would be nice to be able to get only the first valid range for instance for comments.